### PR TITLE
[#137631, #139373] admin hold expiration and UI addition

### PR DESCRIPTION
**Pull request by meara**
https://pm.tablexi.com/issues/139373 - soft delete expired admin reservations
https://pm.tablexi.com/issues/137631 - add end times to admin reservation scheduler

- [x] add expiration info to hover-overs in calendar views
- [x] try out in NU fork (verify things work with oracle)
- [x] fix expiration styling

### CODE REVIEW

File: app/controllers/facility_reservations_controller.rb [hunk 1]
```diff
@@ -141,6 +141,7 @@ def update_admin
 
     @reservation.assign_times_from_params(params[:admin_reservation])
     @reservation.admin_note = params[:admin_reservation][:admin_note]
+    @reservation.expires_mins_before = params[:admin_reservation][:expires_mins_before]
     @reservation.category = params[:admin_reservation][:category]
```
Comment (jhanggi) [inline app/controllers/facility_reservations_controller.rb]: If we can use strong params to make this a little cleaner, I would vote for that.

Comment (jhanggi) [reply app/controllers/facility_reservations_controller.rb]: It's a also a little weird the stylistic difference from `create_admin`. Maybe there's a reason they're different, but if we can make them more similar, that'd be great.

File: config/locales/en.models.yml [hunk 2]
```diff
@@ -330,6 +330,7 @@ en:
         canceled_by: Canceled By
         canceled_at: Canceled At
         type: Type
+        expires_mins_before: Expires Time Prior
```
Comment (jhanggi) [inline config/locales/en.models.yml]: I'm not sure what, but I feel like we should try to come up with a better label for this field. Maybe a hint would help?

File: db/migrate/20171003201727_add_expires_minutes_before_to_reservations.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddExpiresMinutesBeforeToReservations < ActiveRecord::Migration
+  def change
+    add_column :reservations, :expires_mins_before, :integer
```
Comment (jhanggi) [inline db/migrate/20171003201727_add_expires_minutes_before_to_reservations.rb]: We should validate the numericality of this field and that it's only integers. There's a part of me that wants to allow negative numbers (if the timeinput will allow that), although that might be moot with being allowed to schedule over an admin reservation (I'm thinking that time opens up in the past so you can retroactively reserve for people).

File: db/schema.rb [hunk 4]
```diff
@@ -753,7 +755,7 @@
     t.datetime "updated_at",             null: false
   end
 
-  add_index "user_preferences", ["user_id", "name"], name: "index_user_preferences_on_user_id_and_name", unique: true, using: :btree
+  add_index "user_preferences", ["user_id"], name: "index_user_preferences_on_user_id", using: :btree
```
Comment (jhanggi) [inline db/schema.rb]: This seems like it doesn't belong in this PR.

Comment (meara) [reply db/schema.rb]: https://github.com/tablexi/nucore-open/pull/1247

Comment (meara) [reply db/schema.rb]: Also, apparently `expires_mins_before` has already snuck into the schema (in https://github.com/tablexi/nucore-open/pull/1229/files#diff-1acd2e7e27a227829d5d14a91c863bb6R551). Is that going to cause an issue?

Comment (jhanggi) [reply db/schema.rb]: See my comment on #1247, the original is what the schema _should_ be.

The `expires_mins_before` sneaking in shouldn't cause any problems as long as later PRs didn't revert it.

File: db/migrate/20171003201727_add_expires_minutes_before_to_reservations.rb (hunk 3 above)
Comment (meara) [reply db/migrate/20171003201727_add_expires_minutes_before_to_reservations.rb]: Added.

File: config/locales/en.models.yml (hunk 2 above)
Comment (meara) [reply config/locales/en.models.yml]: We'll see what NU thinks.

File: app/helpers/date_helper.rb [hunk 5]
```diff
@@ -56,6 +56,14 @@ def human_time(dt)
     end
   end
 
+  def display_as_time(minutes)
```
Comment (jhanggi) [inline app/helpers/date_helper.rb]: I don't have a problem with adding a view helper method for it, but we have `MinutesToTimeFormatter` which includes this logic.

File: app/presenters/reservations/calendar_presenter.rb [hunk 6]
```diff
@@ -23,6 +23,12 @@ def reservation_options(options)
           title: "Instrument Offline",
           email: created_by.try(:full_name),
         }
+      elsif expires_mins_before.present?
```
Comment (jhanggi) [inline app/presenters/reservations/calendar_presenter.rb]: It's not bad, but I might do something like

```
else
  hash = {
    title: model_name.human,
    email: created_by.try(:full_name),
  }
  hash[:expiration] = "Expires #{display_as_time(expires_mins_before)} prior" if expires_mins_before.present?
  hash
```

We don't necessarily need to do it, but what would happen if we changed `OfflineReservations`'s model name to "Instrument Offline" and collapse even another conditional?

File: app/views/facility_reservations/_admin_reservation_fields.html.haml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      .controls
+        .string.optional.control-label &nbsp;
+        = time_select f, :reserve_end, minute_step: 1
+
+.clearfix
+
+= f.input :admin_note, input_html: { class: "span6" } unless @order_detail
+
+.checkboxControl
+  = f.input :expires?, as: :boolean, input_html: { data: { "disables" => ".admin_reservation_expires_mins_before" } }, wrapper_html: {class: "checkboxControl__checkbox"}
```
Comment (jhanggi) [inline app/views/facility_reservations/_admin_reservation_fields.html.haml]: Some spacing issues inside of `{}`, and can we use the symbol `disables:`?

File: app/views/facility_reservations/edit.html.haml [hunk 8]
```diff
@@ -24,7 +24,38 @@
...
+          .string.optional.control-label &nbsp;
+          = time_select f, :actual_start
+      = f.input :actual_duration_mins, input_html: { class: "timeinput" }
+    - unless @reservation.has_actuals?
+      .copy_actual_from_reservation
+        = link_to "Copy from reservation", "#"
+        %span.start_date.hidden= @reservation.reserve_start_at
+        %span.end_date.hidden= @reservation.reserve_end_at
+    - if @instrument.reservation_only?
+      %li.alert This is a reservation only instrument. Pricing is determined by the reservation times.
```
Comment (jhanggi) [inline app/views/facility_reservations/edit.html.haml]: We probably should i18n this at some point, but doesn't have to be now.

File: app/views/facility_reservations/edit_admin.html.haml [hunk 9]
```diff
@@ -1,5 +1,5 @@
 = content_for :head_content do
-  = render :partial => 'shared/headers/calendar'
+  = render 'shared/headers/calendar'
```
Comment (jhanggi) [inline app/views/facility_reservations/edit_admin.html.haml]: We're cleaning up the hash syntax here. Can we clean up the single v double quotes?

File: app/assets/javascripts/app/disabling_by_checkbox.coffee [hunk 10]
```diff
@@ -0,0 +1,8 @@
+$ ->
+  $("[data-disables]").on "change", ->
```
Comment (jhanggi) [inline app/assets/javascripts/app/disabling_by_checkbox.coffee]: Code climate is complaining about `this` (aka `@`) without a fat arrow. If we make this a fat arrow, I think we would need to add something like `.on "change", (el) ->` and then `$(el)` (I may be wrong on the exact syntax).

File: app/helpers/date_helper.rb (hunk 5 above)
Comment (meara) [reply app/helpers/date_helper.rb]: Ah thank you, I thought we had something like that but couldn't find it.

File: app/assets/javascripts/app/disabling_by_checkbox.coffee (hunk 10 above)
Comment (meara) [reply app/assets/javascripts/app/disabling_by_checkbox.coffee]: Yeah I kinda don't want to mess with the javascript right now if that's okay with you.

File: app/presenters/reservations/calendar_presenter.rb (hunk 6 above)
Comment (meara) [reply app/presenters/reservations/calendar_presenter.rb]: I don't entirely know all the spots where we use OfflineReservation's model name so am a little hesitant to do that now, but you make good sense.

File: app/views/facility_reservations/edit.html.haml (hunk 8 above)
Comment (meara) [reply app/views/facility_reservations/edit.html.haml]: I vote for later. ;)
